prml/generic-asset/src/lib.rs

```diff
@@ -420,20 +420,45 @@ decl_module! {
 
 		/// On runtime upgrade, update account data for existing accounts and remove dust balances
 		fn on_runtime_upgrade() -> frame_support::weights::Weight {
-			let mut dust_accounts = <Vec<(T::AssetId, T::AccountId)>>::new();
-			<FreeBalance<T>>::iter().for_each(|(asset_id, account_id, _)|{
-				let is_dust = Self::is_dust(asset_id, &account_id);
-				Self::update_account_store(asset_id, &account_id, is_dust);
-				if is_dust {
-					dust_accounts.push((asset_id, account_id));
-				}
-			});
-			dust_accounts.iter().for_each(|(asset_id, account_id)| Self::purge(*asset_id, account_id));
+			if StorageVersion::get() == Releases::V0 as u32 {
+				StorageVersion::put(Releases::V1 as u32);
+
+				migrate_locks::<T>();
+
+				let mut dust_accounts = <Vec<(T::AssetId, T::AccountId)>>::new();
+				<FreeBalance<T>>::iter().for_each(|(asset_id, account_id, _)|{
+					let is_dust = Self::is_dust(asset_id, &account_id);
+					Self::update_account_store(asset_id, &account_id, is_dust);
+					if is_dust {
+						dust_accounts.push((asset_id, account_id));
+					}
+				});
+				dust_accounts.iter().for_each(|(asset_id, account_id)| Self::purge(*asset_id, account_id));
```

Comment:
could just purge inline instead of allocating a new vec (kind of worried this whole process will be slow on mainnet)

```suggestion
						Self::purge(*asset_id, account_id);
					}
				});
```

Reply:
Purge internally changes FreeBalance and so not safe to be called while iterating on FreeBalance. We will test the speed on Nikau.